Feature request per pod

therahulbhati opened the pull request:
Addresses request per pod configurable part of #1888

Introduced a parameter `--requestsperpod(--rpp) ` default to 1
 
Max request that can be served concurrently will be equal to concurrency x rpp

Scenarios:
1. The very first request comes a new pod is specialized to serve the request.
2. More request comes, rpp for a single pod is reached, new pod is specialized to serve the request.
3. If all pods are serving requests then return 429.
4. If a pod is serving 'x' requests(x < rpp) and CPU consumed is more than 85 percent then specialize another pod to serve the request.
5. If all pods are serving in total 'y' request (y < maxConcurrentRequests) and all of them have CPU utilization of more than 85 percent, then return 429. 

Note: If fetching of CPU utilization fails, scaling decision would be taken on the only number of requests.

Testing:
With burst for 20 Request for 2 secs |  function rpp 10
1. Warm the cache with 2 requests, this will specialize two functions and add them to cache.
2. Send 20 Requests for 2 secs, all traffic will be served by two already function pods. (watch function pods with label managed=false in fission-function namespace)

therahulbhati commented on pkg/apis/core/v1/types.go:
```diff
@@ -353,6 +353,14 @@ type (
+
+		// RequestsPerPod indicates the maximum number of concurrent requests that can be served by a specialized pod
+		// This is optional. If not specified default value will be taken as 5
+		RequestsPerPod int `json:"requestsperpod,omitempty"`
```
Done

therahulbhati commented on pkg/crd/crdvalidations.go:
```diff
@@ -49,6 +49,14 @@ var (
+				},
+				"requestsperpod": {
+					Type:        "integer",
+					Description: "RequestsPerPod indicates the maximum number of concurrent requests that can be served by a specialized pod.\n This is optional. If not specified default value will be taken as 5",
```
Done

therahulbhati replied:
Done

therahulbhati commented on pkg/router/router.go:
```diff
@@ -123,7 +123,7 @@ func Start(logger *zap.Logger, port int, executorURL string) {
 	fmap := makeFunctionServiceMap(logger, time.Minute)
 
-	fissionClient, kubeClient, _, err := crd.MakeFissionClient()
+	fissionClient, kubeClient, _, _, err := crd.MakeFissionClient()
```
Yes, its in progess

therahulbhati commented on pkg/crd/client.go:
```diff
@@ -51,45 +52,47 @@ func GetKubernetesClient() (*rest.Config, *kubernetes.Clientset, *apiextensionsc
 	}
 
-	return config, clientset, apiExtClientset, nil
+	metricsClient, _ := metricsclient.NewForConfig(config)
```
If metric server is not installed, then CPU aspect of scaling would be ignored and it work as aspected. However, there would error message for unable to metrics API call failed like `failed to fetch pod metrics list	{"error": "the server could not find the requested resource (get pods.metrics.k8s.io)"}
`

therahulbhati commented on pkg/fission-cli/cmd/function/create.go:
```diff
@@ -102,6 +102,10 @@ func (opts *CreateSubCommand) complete(input cli.Input) error {
 		fnConcurrency = input.Int(flagkey.FnConcurrency)
 	}
 
+	fnOnceOnly := input.Bool(flagkey.FnOnceOnly)
```
removed `onceonly` from this PR

therahulbhati commented on pkg/executor/executortype/poolmgr/gp.go:
```diff
@@ -161,6 +166,22 @@ func (gp *GenericPool) getDeployAnnotations() map[string]string {
+		gp.logger.Error("failed to fetch pod metrics list", zap.Error(err))
+	} else {
+		for item := range podMetricsList.Items {
+			gp.logger.Info(fmt.Sprintf("%+v", item))
```
yes

therahulbhati commented on pkg/executor/executortype/poolmgr/gp.go:
```diff
@@ -161,6 +166,22 @@ func (gp *GenericPool) getDeployAnnotations() map[string]string {
 
+func (gp *GenericPool) updateCPUUtilizationSvc() {
+	podMetricsList, err := gp.metricsClient.MetricsV1beta1().PodMetricses(gp.env.ObjectMeta.Namespace).List(context.Background(), v1.ListOptions{
+		LabelSelector: labels.Set(gp.deployment.Spec.Selector.MatchLabels).AsSelector().String(),
```
Done

vishal-biyani commented:
@therahulbhati Can you please share how you did the testing for this feature. I tried with an RPP of 4 and added sleep to the function for about 8 seconds, but even then when I send new requests - it specialized new pods instead of reusing the same one. 

Also it would be good to add some sort of tests to the PR.

vishal-biyani commented on pkg/executor/api.go:
```diff
@@ -50,25 +50,59 @@ func (executor *Executor) getServiceForFunctionAPI(w http.ResponseWriter, r *htt
+				executor.writeResponse(w, fsvc.Address, fn.ObjectMeta.Name)
+				return
+			}
+			executor.logger.Debug("deleting cache entry for invalid address",
+				zap.String("function_name", fn.ObjectMeta.Name),
+				zap.String("function_namespace", fn.ObjectMeta.Namespace),
+				zap.String("address", fsvc.Address))
+			et.DeleteFuncSvcFromCache(fsvc)
+			active--
```
Should this not go in an else block for the `isValid` condition? Otherwise, this will be executed even if a service `isValid` by default and when active <= 1?

therahulbhati replied:
In '||' condition if the first condition is evaluated to true, subsequent conditions aren't evaluated.

vishal-biyani commented on pkg/executor/api.go:
```diff
@@ -50,25 +50,59 @@ func (executor *Executor) getServiceForFunctionAPI(w http.ResponseWriter, r *htt
+			active--
+		}
+
+		if active >= concurrency {
```
This block will be executed only if the `err == nil` condition is not met but this will only be clear if you actually dig down the stack and see in implementation. I was confused while reading - so suggest adding at least a comment to that effect!

therahulbhati replied:
Added

vishal-biyani commented on pkg/poolcache/poolcache.go:
```diff
@@ -52,13 +56,15 @@ type (
 		address         interface{}
 		value           interface{}
+		requestsPerPod  int
+		cpuUsage        resource.Quantity
```
`cpuUsage` is actual usage or a place to store limit for that function?

therahulbhati replied:
Its used for updating both CPU usage and CPU limit

vishal-biyani commented on pkg/poolcache/poolcache.go:
```diff
@@ -22,6 +22,7 @@ import (
 	ferror "github.com/fission/fission/pkg/error"
+	"k8s.io/apimachinery/pkg/api/resource"
 )
 
```
This cache's structure is fairly complex at this point, would be nice to add a high level comment with structure and usage so it is easier on reader

therahulbhati replied:
Done

vishal-biyani reviewed (changes requested):
Hey, @therahulbhati Some comments and I think it would be good to add overall tests. Also, some words on how testing was done will be useful. ATM I am not able to make it work as commented in the previous comment

RealHarshThakur commented on pkg/mqtrigger/scalermanager_test.go:
```diff
@@ -386,40 +382,44 @@ func newUnstructured(apiVersion, kind, namespace, name, resourceVersion string)
-		})
-	}
-}
+// commented because this fails with k8s.io/client-go v0.17.2
```
What's the error?

RealHarshThakur commented:
@therahulbhati what happens when there are 3 pods with RPP of function is set to 5. All of them are serving to max capacity(15 in total), the first pod completes two requests. Would the next two requests arrive at the first pod or a new pod would be spun up? I couldn't figure out which part of code this decision making happens

therahulbhati commented:
> @therahulbhati Can you please share how you did the testing for this feature? I tried with an RPP of 4 and added sleep to the function for about 8 seconds, but even then when I send new requests - it specialized new pods instead of reusing the same one.
> 
> Also, it would be good to add some sort of tests to the PR.

Updated the description

therahulbhati commented on pkg/executor/executortype/poolmgr/gp.go:
```diff
@@ -632,17 +682,43 @@ func (gp *GenericPool) getFuncSvc(ctx context.Context, fn *fv1.Function) (*fscac
 }
 
+// get85Percent returns 85 percent of the quantity i.e multiple it by 17/20
+func (gp *GenericPool) get85Percent(cpuUsage resource.Quantity) (resource.Quantity, error) {
```
@RealHarshThakur the method you suggested in this I would have to convert and parse suffix like milli, nano etc which would be tedious. 
@vishal-biyani I think this will be useful, this would make code cleaner.

therahulbhati commented on pkg/executor/executortype/executortype.go:
```diff
@@ -38,6 +38,9 @@ type ExecutorType interface {
 	GetFuncSvcFromCache(*fv1.Function) (*fscache.FuncSvc, error)
 
+	// GetFuncSvcFromPoolCache retrieves function service after filtering on requestsPerPod and CPULimit
+	GetFuncSvcFromPoolCache(fn *fv1.Function, requestsPerPod int) (*fscache.FuncSvc, int, error)
```
No, 0 makes more sense when specialized function pod is serving any requests.

therahulbhati commented:
> @therahulbhati what happens when there are 3 pods when RPP of function is set to 5 and all of them are serving to max capacity(15 in total), the first pod completes two requests. Would the next two requests arrive at the first pod or a new pod would be spun up? I couldn't figure out which part of code this decision making happens

Yes @RealHarshThakur two requests would be served by the first pod. The logic is here https://github.com/fission/fission/pull/1946/files#diff-35348f6046f2853651e6b55a3508c7efe61e0849768957cb193f8d4e63668aa3R94

vishal-biyani reviewed (approved):
I was able to test this. Except for initial cache misses if too many requests come at the same time - this PR looks good. I think it would be worth investing in doing a separate optimization around how to improve initial cache miss and other cache problems!

RealHarshThakur commented on pkg/executor/executortype/poolmgr/gp.go:
```diff
@@ -623,6 +661,18 @@ func (gp *GenericPool) getFuncSvc(ctx context.Context, fn *fv1.Function) (*fscac
+	cpuLimit, err := gp.get85Percent(cpuUsage)
+	if err != nil {
+		gp.logger.Error("failed to get 85 of CPU usage", zap.Error(err))
+		cpuLimit = cpuUsage
```
Then instead of setting to current cpu Usage which could be high(90% +), we could set it to 85% . What do you think? @therahulbhati

therahulbhati replied:
That cpuUsage that we are setting is from Resources.Limits.Cpu of Pod

vishal-biyani commented:
Thanks for landing this @therahulbhati 🎉
